examples/gameroom/daemon/src/rest_api/routes/authenticate.rs

```diff
@@ -32,28 +26,151 @@ pub struct AuthResponseData {
     encrypted_private_key: String,
 }
 
-#[derive(Debug, Deserialize)]
+#[derive(Debug, Deserialize, Serialize)]
 #[serde(rename_all = "camelCase")]
 pub struct AuthData {
     pub email: String,
     pub hashed_password: String,
 }
 
+#[derive(Debug, Deserialize, Serialize)]
+pub struct UsernamePassword {
+    pub username: String,
+    pub hashed_password: String,
+}
+
+impl From<AuthData> for UsernamePassword {
+    fn from(auth_data: AuthData) -> Self {
+        Self {
+            username: auth_data.email,
+            hashed_password: auth_data.hashed_password,
+        }
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+struct Token {
+    pub message: String,
+    pub user_id: String,
+    pub token: String,
+}
+
+#[derive(Clone, Deserialize)]
+struct Keys {
+    pub data: Vec<Key>,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+struct Key {
+    pub public_key: String,
+    pub encrypted_private_key: String,
+    pub user_id: String,
+    pub display_name: String,
+}
+
+#[derive(Serialize)]
+struct NewKey {
+    pub public_key: String,
+    pub encrypted_private_key: String,
+    pub display_name: String,
+}
+
 pub async fn login(
     auth_data: web::Json<AuthData>,
-    pool: web::Data<ConnectionPool>,
+    splinterd_url: web::Data<String>,
+    client: web::Data<Client>,
 ) -> Result<HttpResponse, Error> {
-    match web::block(move || authenticate_user(pool, auth_data.into_inner())).await {
-        Ok(user) => Ok(HttpResponse::Ok().json(SuccessResponse::new(user))),
-        Err(err) => match err {
-            error::BlockingError::Error(_) => Ok(HttpResponse::Unauthorized()
-                .json(ErrorResponse::unauthorized("Invalid email or password"))),
-            error::BlockingError::Canceled => {
-                debug!("Internal Server Error: {}", err);
-                Ok(HttpResponse::InternalServerError().json(ErrorResponse::internal_error()))
+    // forward login to splinterd
+    let mut login_response = client
+        .post(format!("{}/biome/login", splinterd_url.get_ref()))
+        .header(
+            "SplinterProtocolVersion",
+            protocol::ADMIN_PROTOCOL_VERSION.to_string(),
```

Comment:
This should be `BIOME_PROTOCOL_VERSION` (also elsewhere in this file).

Reply:
updated